rbtree_bench_test.go

```diff
@@ -0,0 +1,59 @@
+package rbtree
+
+import (
+	"crypto/rand"
+	"fmt"
+	"math"
+	"math/big"
+	"testing"
+)
+
+func BenchmarkRBTree(b *testing.B) {
+	max := big.NewInt(math.MaxInt)
+	iterations := b.N * 500_000
+
+	keys := make([]int64, iterations)
+	for i := 0; i < iterations; i++ {
+		n, _ := rand.Int(rand.Reader, max)
+		keys[i] = n.Int64()
+	}
+	values := make([]string, iterations)
+	for i := 0; i < iterations; i++ {
+		values[i] = fmt.Sprintf("%20.20d", keys[i])
+	}
+
+	tree := NewTree[int64, string]()
+	b.Run("insert", func(b *testing.B) {
+		b.ReportAllocs()
+
+		for i := 0; i < iterations; i++ {
```

Comment:
Is this correct? Using a fixed-length iterations? As far as i know, we should use b.N in a subbenchmark because b.N changes every time.
If I'm mistaken, please correct me.

Reply:
`iterations` is a multiple of `b.N` (see line 13), but the initial few `b.N` values are too quick so Go doesn't run higher values of `b.N`. We set a higher b.N value to force longer runs.

Reply:
Another option would be to run with a higher multiple than `-benchtime=1_000_000_000x`, but that doesn't seem to work. I'm not aware of a way to force benchmarks to run for higher `b.N` values.

Reply:
In [go doc](https://pkg.go.dev/testing#B.Run):
> A benchmark that calls Run at least once will not be measured itself and will be called once with N=1.

In BenchmarkRBTree, `b.N` will not change. However, in a subbenchmark, `b.N` is always subject to change.

Therefore, it is necessary to use `b.N` as the number of iterations in subbenchmarks.

Reply:
We can setup datas in every subbenchmark, then call `b.ResetTimer()`.

After doing this, the iterations will be exactly the same as b.N, and we can get the most accurate test results.

```go
	b.Run("iterate", func(b *testing.B) {
		fmt.Println(b.N)

		b.ReportAllocs()
		keys := make([]int64, b.N)
		iterations := b.N
		for i := 0; i < iterations; i++ {
			n, _ := rand.Int(rand.Reader, max)
			keys[i] = n.Int64()
		}
		values := make([]string, iterations)
		for i := 0; i < iterations; i++ {
			values[i] = fmt.Sprintf("%20.20d", keys[i])
		}
		tree := NewTree[int64, string]()
		for i := 0; i < iterations; i++ {
			tree.Insert(keys[i], values[i])
		}

		b.ResetTimer()
		it := tree.Iterator()
		for it != nil {
			it = it.Next()
		}
	})
```

Reply:
You're right. Updated.